protocols/mdns/src/behaviour/timer.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    marker::Unpin,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+/// Simple wrapper for the differents type of timers
+#[derive(Debug)]
+pub struct WrapTimer<T> {
+    timer: T,
+}
+
+/// Builder interface to homogenize the differents implementations
+pub trait TimerBuilder: Send + Unpin + 'static {
+    type Item;
+
+    /// Creates a timer that emits an event once at the given time instant.
+    fn at(instant: Instant) -> Self;
+
+    /// Creates a timer that emits events periodically.
+    fn interval(duration: Duration) -> Self;
+
+    /// Creates a timer that emits events periodically, starting at start.
+    fn interval_at(start: Instant, duration: Duration) -> Self;
+
+    /// Poll the timer
+    fn poll_tick(&mut self, cx: &mut Context) -> Poll<Self::Item>;
```

Comment:
Instead of this fn here, what about implementing `Stream`?

This would make the trait naming better as this is the only none-builder function atm.